Library/Homebrew/utils/pypi.rb

```diff
@@ -278,20 +278,13 @@ def self.update_python_resources!(formula, version: nil, package_name: nil, extr
 
     ensure_formula_installed!("python")
 
+    # Resolve the dependency tree of all input packages
     ohai "Retrieving PyPI dependencies for \"#{input_packages.join(" ")}\"..." if !print_only && !silent
-    command =
-      [Formula["python"].bin/"python3", "-m", "pip", "install", "-q", "--dry-run", "--ignore-installed", "--report",
-       "/dev/stdout", *input_packages.map(&:to_s)]
-    pip_output = Utils.popen_read({ "PIP_REQUIRE_VIRTUALENV" => "false" }, *command)
-    unless $CHILD_STATUS.success?
-      odie <<~EOS
-        Unable to determine dependencies for "#{input_packages.join(" ")}" because of a failure when running
-        `#{command.join(" ")}`.
-        Please update the resources for "#{formula.name}" manually.
-      EOS
-    end
-
-    found_packages = pip_report_to_packages(JSON.parse(pip_output), exclude_packages).uniq
+    found_packages = pip_report(input_packages)
+    # Resolve the dependency tree of excluded packages to prune the above
+    exclude_packages.delete_if { |package| found_packages.exclude? package }
+    ohai "Retrieving PyPI dependencies for excluded \"#{exclude_packages.join(" ")}\"..." if !print_only && !silent
+    exclude_packages = pip_report(exclude_packages) + [Package.new(main_package.name)]
```

Comment:
This might incorrectly exclude common deps shared by the main package and some of the `exclude_packages`, I think.

Take https://github.com/Homebrew/homebrew-core/pull/140275 as an example (see https://github.com/Homebrew/homebrew-core/commit/f4c55eacb8bf780c6dd1e4d7ce667f816aacace1). Here, `moto`, the main package, depends on `boto3` and `botocore`. One of its `exclude_packages`, `cfn-lint`, also depends on `boto3` and `botocore`. IIUC, with this logic, `boto3` and `botocore` are incorrectly removed.

Reply:
I think that's what we want, since in python there's only a single version of a dep shared by the whole venv (so no need to install it twice).

That formula actually has a bug, its depending on `cfn-lint` without it being importable so neither is `boto3` (it needs a `.pth` file to find it)

```
$(brew --prefix)/opt/moto/libexec/bin/python -c 'import cfnlint'
Traceback (most recent call last):
  File "<string>", line 1, in <module>
ModuleNotFoundError: No module named 'cfnlint'
```